### PR TITLE
Ac 2162 form UI customisation

### PR DESCRIPTION
**Pull request by soeren-schmaljohann-2denker**

### CODE REVIEW

File: packages/apptive_grid_form/lib/src/apptive_grid_form.dart [hunk 1]
```diff
@@ -492,10 +494,18 @@ class ApptiveGridFormDataState extends State<ApptiveGridFormData> {
...
           textAlign: TextAlign.center,
           style: Theme.of(context).textTheme.headline4,
         ),
+        if (_formData?.successMessage != null) ...[
+          const SizedBox(height: 4),
+          Text(
+            _formData!.successMessage!,
+            textAlign: TextAlign.center,
+            style: Theme.of(context).textTheme.bodyMedium,
+          ),
```
Comment (ABausG) [inline packages/apptive_grid_form/lib/src/apptive_grid_form.dart:507]: How is the web doing this? Is there now a default message? In that case it should obviously be used instead

File: packages/apptive_grid_form/lib/src/apptive_grid_form.dart [hunk 2]
```diff
@@ -492,10 +494,18 @@ class ApptiveGridFormDataState extends State<ApptiveGridFormData> {
...
+        ],
         Center(
           child: TextButton(
```
Comment (ABausG) [inline packages/apptive_grid_form/lib/src/apptive_grid_form.dart:510]: Would it make sense to allow customisation of this text similar to how the package allows for a custom send button text?

File: packages/apptive_grid_form/test/apptive_grid_form_test.dart [hunk 3]
```diff
@@ -332,6 +371,43 @@ void main() {
...
+          SubmitCompleteProgressEvent(http.Response('', 200)),
+        ),
+      );
+
+      await tester.pumpWidget(target);
+      await tester.pumpAndSettle();
+      await tester.tap(find.byType(ActionButton));
+      await tester.pumpAndSettle();
+
+      verify(
```
Comment (ABausG) [inline packages/apptive_grid_form/test/apptive_grid_form_test.dart:407]: This should also check that that the success state is not shown but an empty form is shown

**Review by ABausG — COMMENTED** [review]
Some minor tweaks, questions regarding the redirect stuff from the ticket and questions regarding customising the add another entry button in the success state

File: packages/apptive_grid_core/CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,10 @@
 ## 0.13.1
 * Parse multiline text to single line for single line form fields
+* Add new properties to forms
+  * successTitle
+  * successMessage
+  * buttonTitle
+  * reloadAfterSubmit
```
[the remark this answers is not in the document]
Comment (soeren-schmaljohann-2denker) [reply packages/apptive_grid_core/CHANGELOG.md:7]: This is Web-URL (at least as far as I can tell), so it doesn't make sense to open a web page after a successful form submission, when you are in a mobile app.

File: packages/apptive_grid_form/lib/src/apptive_grid_form.dart (hunk 2 above)
Comment (soeren-schmaljohann-2denker) [reply packages/apptive_grid_form/lib/src/apptive_grid_form.dart:510]: Seems like something we could do as a nice to have feature, but maybe not in this PR. I'll make a ticket for it

Comment (soeren-schmaljohann-2denker) [reply packages/apptive_grid_form/lib/src/apptive_grid_form.dart:510]: https://2denker.atlassian.net/browse/AC-2325

File: packages/apptive_grid_form/lib/src/apptive_grid_form.dart (hunk 1 above)
Comment (soeren-schmaljohann-2denker) [reply packages/apptive_grid_form/lib/src/apptive_grid_form.dart:507]: Currently the web doesn't have a default message here

File: packages/apptive_grid_core/lib/src/model/form/form_data.dart [hunk 5]
```diff
@@ -15,6 +15,10 @@ class FormData {
     required this.fields,
     required this.links,
     Map<Attachment, AttachmentAction>? attachmentActions,
+    this.buttonTitle,
+    this.reloadAfterSubmit,
+    this.successTitle,
+    this.successMessage,
```
[the remark this answers is not in the document]
Comment (soeren-schmaljohann-2denker) [reply packages/apptive_grid_core/lib/src/model/form/form_data.dart]: done

File: packages/apptive_grid_form/test/apptive_grid_form_test.dart (hunk 3 above)
Comment (soeren-schmaljohann-2denker) [reply packages/apptive_grid_form/test/apptive_grid_form_test.dart:412]: done

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -29,7 +29,7 @@ jobs:
         uses: VeryGoodOpenSource/very_good_coverage@v1
         with:
           path: ./packages/apptive_grid_core/coverage/lcov.info
-          min_coverage: 99.6
+          min_coverage: 99.2
```
Comment (ABausG) [inline .github/workflows/main.yml:32]: Which lines are not hit? Is it misreported lines or are there just tests missing?

Comment (soeren-schmaljohann-2denker) [reply .github/workflows/main.yml:32]: It's the hash generator, that is not hit

Comment (ABausG) [reply .github/workflows/main.yml:32]: Then please add a test for the hash function. You can look at other model tests for reference

File: packages/apptive_grid_core/test/form_data_test.dart [hunk 7]
```diff
@@ -323,6 +355,10 @@ void main() {
     test('a != c', () {
       expect(a, isNot(c));
     });
+
+    test('properties hash', () {
+      expect(a.properties.hashCode, b.properties.hashCode);
+    });
```
Comment (ABausG) [inline packages/apptive_grid_core/test/form_data_test.dart:361]: I rather meant that you'd add tests for equality, hashcode and the toString method for the `FormDataProperties` as a new testy group for the properties.

For reference here is testcode for the `CurrencyDataEntity` from `currency_test.dart`

```
  group('DataEntity', () {
    test('Equality', () {
      final a = CurrencyDataEntity(currency: 'USD', value: 12345.12);
      final b = CurrencyDataEntity(currency: 'USD', value: 12345.12);
      final c = CurrencyDataEntity(currency: 'EUR', value: 12345.12);
      expect(a, equals(b));
      expect(a, isNot(c));
    });

    test('Hashcode', () {
      final currency = CurrencyDataEntity(currency: 'USD', value: 12345.12);

      expect(currency.hashCode, Object.hash(12345.12, 'USD'));
    });

    test('toString()', () {
      final currency = CurrencyDataEntity(currency: 'USD', value: 12345.12);

      expect(currency.toString(), equals('CurrencyDataEntity(12345.12 USD)'));
    });
  });
```

**Review by ABausG — APPROVED** [review]

**Pull request merged**